### ISSUE TITLE
Measuring kubernetes pods

### ISSUE THREAD

**Issue by catherinemeadows** [powerapi-ng/smartwatts-formula#28]
Hello, 

I am attempting to use the HWPC sensor + SmartWatts formula to measure the energy consumption of individual pods in my kubernetes cluster. I've been struggling to find documentation on how to point the sensor / formula to get container/pod-level measurements. Running the default configuration from the documentation only gives system wide measurements. When I tried adding pids to a cgroup and then specifying a "target" with the cgroup name in the sensor config file, the HWPC sensor failed to start up saying it was an invalid json file. 

Any direction would be very helpful, thank you!

**Comment by roda82** [comment]
Hello, 

Do you create your cgroup with 'perf_event:' as prefix ? Are you sure that your PID is actually associated to your cgroup ? You cannot indicate a target element in the sensor configuration file. If your PID is correctly associated with your 'perf_event' cgroup, in the ouput database you should found estimations for your container by using the target as filter in your query.

**Comment by catherinemeadows** [comment]
Hi, 

I'm using a 2 node configuration with the HWPC sensor running on the worker nod  and the smartwatts formula running on the control plane node. The HWPC sensor inputs to a mongodb on the control plane node. 

The steps I followed for each worker node are: 
1. Make a cgroup with `sudo cgcreate -g perf_event:<name>`
2. Added the PIDs of the pods on the worker node with `sudo cgclassify -g perf_event:<name> <PID>`
3. Run HWPC on worker node. Config file:
```
{
    "name": "sensor",
    "verbose": true,
    "frequency": 500,
    "output": {
      "type": "mongodb",
      "uri": "mongodb://10.10.1.1",
      "database": "from_node_1",
      "collection": "sensor_data"
    },
    "system": {
      "rapl": {
        "events": ["RAPL_ENERGY_PKG"],
        "monitoring_type": "MONITOR_ONE_CPU_PER_SOCKET"
      },
      "msr": {
        "events": ["TSC", "APERF", "MPERF"]
      }
    },
    "container": {
      "core": {
        "events": [
          "CPU_CLK_THREAD_UNHALTED:REF_P",
          "CPU_CLK_THREAD_UNHALTED:THREAD_P",
          "LLC_MISSES",
          "INSTRUCTIONS_RETIRED"
        ]
      }
    }
  }
```
4. Run smart watts formula on control node
Config file: 
```
{
    "verbose": true,
    "stream": true,
    "input": {
      "puller": {
        "model": "HWPCReport",
        "type": "mongodb",
        "uri": "mongodb://127.0.0.1",
        "db": "from_node_1",
        "collection": "sensor_data"
      }
    },
    "output": {
      "pusher_power": {
        "type": "mongodb",
        "uri": "mongodb://127.0.0.1",
        "db": "formula_data",
        "collection": "formula_data"
      }
    },
    "cpu-frequency-base": 19,
    "cpu-frequency-min": 4,
    "cpu-frequency-max": 42,
    "cpu-error-threshold": 2.0,
    "disable-dram-formula": true,
    "sensor-report-sampling-interval": 1000
  }
```
5. Looking at the formula db, I always get a power measurement of 0 even though I'm sending requests to load test the pods of the cgroup (see image). For the system wide measurements, it reports energy measurements > 0. 
<img width="700" alt="Screen Shot 2022-11-07 at 12 11 53 PM" src="https://user-images.githubusercontent.com/43352540/200372669-e887b78a-4b6f-42c0-b011-1104cb170d37.png">

**Comment by roda82** [comment]
Hi, 
Thank you for the detailed description. Are the worker nod and pods running on the same physical server ? If is no the case, you have to deploy a sensor instance on each physical machine used to run a pod, create the related cgroups and associated them with PIDs of your pods. 

Regarding your configuration files, `sensor-report-sampling-interval` (in smartwatts formula configuration) and `frequency` (in HWPC sensor configuration) must have the same value.

**Comment by catherinemeadows** [comment]
Hi, 

I amended the `sensor-report-sampling-interval` so they both have the same value and am still getting the same error of always getting a value of 0 for the monitored cgroup. I am sending requests to the pods successfully with the hey load generator tool. 

Yes, the pods are running on the worker node which is on a different physical service than the control plane node. I attached an image of my setup.  
<img width="1127" alt="Screen Shot 2022-11-08 at 10 07 18 AM" src="https://user-images.githubusercontent.com/43352540/200600641-4d5021a6-724b-4437-9c31-f9f50432c0ab.png">
